src/py2spack/loading.py

```diff
@@ -0,0 +1,266 @@
+
+
+def _acceptable_version(version: str) -> Optional[vn.Version]:
+    """Check whether version string can be parsed and has correct format."""
+    try:
+        v = vn.parse(version)
+        # do not support post releases of prereleases etc.
+        if v.pre and (v.post or v.dev or v.local):
+            return None
+        return v
+    except vn.InvalidVersion:
+        return None
+
+
+class PyPILookup:
+    """Performs PyPI API calls and caches package versions.
+
+    The full API response (including available files and metadata) is not cached
+    due to its larger size and sparser usage.
+    """
+
+    def __init__(self):
+        """Initialize empty PyPILookup."""
+        # cache for package name and corresponding version list
+        self.version_cache: Dict[str, List[vn.Version]] = {}
+
+    def _get(self, name: str) -> dict | APIError:
+        """Load info for the available distribution files from PyPI.
+
+        Available versions are cached.
+        """
+        r = requests.get(
+            f"https://pypi.org/simple/{name}/",
+            headers={"Accept": "application/vnd.pypi.simple.v1+json"},
+            timeout=10,
+        )
+        if r.status_code != 200:
+            if r.status_code == 404:
+                msg = f"Package {name} not found on PyPI (status code 404)"
+                return APIError(msg)
+
+            msg = (
+                f"Error when querying JSON API (status code {r.status_code})."
+                f" Response: {r.text}"
+            )
+            return APIError(msg)
+
+        data: dict = r.json()
+        versions = data["versions"]
+
+        # parse and sort versions
+        parsed_versions = sorted(
+            {vv for v in versions if (vv := _acceptable_version(v))}
+        )
+
+        self.version_cache[name] = parsed_versions
+
+        return data
+
+    def get_versions(self, name: str) -> List[vn.Version]:
+        """Get acceptable versions for package (from cache if possible)."""
+        if name not in self.version_cache:
+            data = self._get(name)
+
+            if isinstance(data, APIError):
+                self.version_cache[name] = []
+
+        return self.version_cache[name]
+
+    def get_files(
+        self, name: str, last_n_versions=-1
+    ) -> List[Dict[Any, Any]] | APIError:
+        """Get metadata for available distribution files from PyPI.
+
+        Returns:
+            - List of dictionaries including filename, download url, version,
+            sdist archive extension, and archive sha256 hash.
+            If an error occurs, returns APIError object.
+        """
+        data = self._get(name)
+        if isinstance(data, APIError):
+            return data
+        files = data["files"]
+
+        # for now we only support tarball archives like .tar.gz
+        files_known_format = [
+            f for f in files if _known_archive_format(f["filename"])
+        ]
+
+        if len(files_known_format) == 0:
```

Comment:
The idiomatic Python way of checking if builtin collections are empty is check if they are false values:
```suggestion
        if not files_known_format:
```
Also: https://google.github.io/styleguide/pyguide.html#214-truefalse-evaluations